### PR TITLE
Added setup for dummy PyArrow sdists to inform users to load the Arro…

### PR DESCRIPTION
**Pull request by ccoulombe**
…w module

When `pip` picks up the `sdist`, it tries to install it and it raises an exception with the message:
```
              This is a normal error generated by this dummy wheel.
      
      
              PyArrow is available from the Arrow module, please see https://docs.alliancecan.ca/wiki/Arrow
      
              IMPORTANT: the module must be loaded before activating your virtual environment.
      
              1. Deactivate your virtual environment : deactivate
              2. Load the Arrow module : module load gcc arrow/x.y.z
              3. Activate your virtual env. : source <env>/bin/activate
              4. And re-run your pip install command.
```

### No version specified
```
(6445) pyarrow (main *) 2023 $ pip install datasets --find-links=dist
Looking in links: /cvmfs/soft.computecanada.ca/custom/python/wheelhouse/gentoo2023/x86-64-v3, /cvmfs/soft.computecanada.ca/custom/python/wheelhouse/gentoo2023/generic, /cvmfs/soft.computecanada.ca/custom/python/wheelhouse/generic, dist
Processing /cvmfs/soft.computecanada.ca/custom/python/wheelhouse/generic/datasets-2.20.0+computecanada-py3-none-any.whl
Collecting filelock (from datasets)
  Using cached filelock-3.15.4-py3-none-any.whl.metadata (2.9 kB)
Processing /cvmfs/soft.computecanada.ca/custom/python/wheelhouse/gentoo2023/generic/numpy-1.26.4+computecanada-cp311-cp311-linux_x86_64.whl (from datasets)
Processing ./dist/pyarrow-9999+dummy.computecanada.tar.gz (from datasets)
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [19 lines of output]
      
      
      
      
              This is a normal error generated by this dummy wheel.
      
      
              PyArrow is available from the Arrow module, please see https://docs.alliancecan.ca/wiki/Arrow
      
              IMPORTANT: the module must be loaded before activating your virtual environment.
      
              1. Deactivate your virtual environment : deactivate
              2. Load the Arrow module : module load gcc arrow/x.y.z
              3. Activate your virtual env. : source <env>/bin/activate
              4. And re-run your pip install command.
      
      
      
      
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed
```

### Version specified
```
(6445) pyarrow (main *) 2023 $ pip install pyarrow==17.0.0 datasets --find-links=dist
Looking in links: /cvmfs/soft.computecanada.ca/custom/python/wheelhouse/gentoo2023/x86-64-v3, /cvmfs/soft.computecanada.ca/custom/python/wheelhouse/gentoo2023/generic, /cvmfs/soft.computecanada.ca/custom/python/wheelhouse/generic, dist
Processing ./dist/pyarrow-17.0.0+dummy.computecanada.tar.gz
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error
```

### With module loaded
```
(10063) pyarrow (main *) 2023 $ ml

Currently Loaded Modules:
  1) CCconfig          3) imkl/2023.2.0 (math)   5) mii/1.1.2           7) gcc/12.3    (t)   9) ucx/1.14.1        11) pmix/4.2.4  13) openmpi/4.1.5   (m)  15) ipykernel/2024a           17) arrow/17.0.0  (t)
  2) gentoo/2023 (S)   4) StdEnv/2023   (S)      6) gcccore/.12.3 (H)   8) hwloc/2.9.1      10) libfabric/1.18.0  12) ucc/1.2.0   14) flexiblas/3.3.1      16) scipy-stack/2024a (math)  18) python/3.11.5 (t)


(10063) pyarrow (main *) 2023 $ pip install datasets --find-links=dist
Looking in links: /cvmfs/soft.computecanada.ca/custom/python/wheelhouse/gentoo2023/x86-64-v3, /cvmfs/soft.computecanada.ca/custom/python/wheelhouse/gentoo2023/generic, /cvmfs/soft.computecanada.ca/custom/python/wheelhouse/generic, dist
Processing /cvmfs/soft.computecanada.ca/custom/python/wheelhouse/generic/datasets-2.20.0+computecanada-py3-none-any.whl
Collecting filelock (from datasets)
  Using cached filelock-3.15.4-py3-none-any.whl.metadata (2.9 kB)
Requirement already satisfied: numpy>=1.17 in /cvmfs/soft.computecanada.ca/easybuild/software/2023/x86-64-v3/Compiler/gcccore/scipy-stack/2024a/lib/python3.11/site-packages (from datasets) (1.26.4+computecanada)
Requirement already satisfied: pyarrow>=15.0.0 in /cvmfs/soft.computecanada.ca/easybuild/software/2023/x86-64-v3/Compiler/gcccore/arrow/17.0.0/lib/python3.11/site-packages (from datasets) (17.0.0)

```

### Dummies
```
pyarrow-1.0.0+dummy.computecanada.tar.gz
pyarrow-10.0.1+dummy.computecanada.tar.gz
pyarrow-11.0.0+dummy.computecanada.tar.gz
pyarrow-12.0.1+dummy.computecanada.tar.gz
pyarrow-13.0.0+dummy.computecanada.tar.gz
pyarrow-14.0.0+dummy.computecanada.tar.gz
pyarrow-14.0.1+dummy.computecanada.tar.gz
pyarrow-15.0.1+dummy.computecanada.tar.gz
pyarrow-16.1.0+dummy.computecanada.tar.gz
pyarrow-17.0.0+dummy.computecanada.tar.gz
pyarrow-2.0.0+dummy.computecanada.tar.gz
pyarrow-4.0.0+dummy.computecanada.tar.gz
pyarrow-5.0.0+dummy.computecanada.tar.gz
pyarrow-7.0.0+dummy.computecanada.tar.gz
pyarrow-8.0.0+dummy.computecanada.tar.gz
pyarrow-9.0.0+dummy.computecanada.tar.gz
pyarrow-9999+dummy.computecanada.tar.gz
```

### CODE REVIEW

File: dummy/pyarrow/setup.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
+import os,sys
+from setuptools import setup
+
+
+class bcolors:
```
Comment (mboisson) [inline dummy/pyarrow/setup.py:5]: il doit sûrement avoir une classe qui existe déjà pour ça ?

Comment (mboisson) [reply dummy/pyarrow/setup.py:5]: https://pypi.org/project/termcolor/

Comment (ccoulombe) [reply dummy/pyarrow/setup.py:5]: Non sauf si je dépend sur un package externe comme `termcolor` ou `rich`.

File: dummy/pyarrow/README.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+python -c 'import pyarrow'
+virtualenv --no-download --clear ~/ENV && source ~/ENV/bin/activate
+pip install --no-index ...
+```
+
+# Dev
+Generate `sdist` in `dist` directory.
+```bash
+python setup.py sdist; # default higher version
+for v in 17.0.0 16.0.1 16.0.0 15.0.2 15.0.1 15.0.0; do # define specific versions
```
[the remark this answers is not in the document]
Comment (ccoulombe) [reply dummy/pyarrow/README.md:20]: C'était surtout de documenter comment faire pour générer le sdist

[the remark this answers is not in the document]
Comment (mboisson) [reply dummy/pyarrow/README.md:20]: ok

**Comment by mboisson** [comment]
Are there a lot of minor versions and revisions of PyArrow ? 

If so, if, say, 16.0.2 comes out and something requires `~=16`, or `<17`, it would install that version. Shouldn't we code subversions as `16.999` for example ?

**Comment by ccoulombe** [comment]
There's not many releases of Arrow/PyArrow : https://pypi.org/project/pyarrow/#history

1 or 2 minor releases so far per major releases.

Since we don't have a module (that version), it would still fail at installing the version of pyarrow 16.0.2 and the user would end up opening a ticket.

We could indeed add a revision of `.999` for the current major release.
For whats it's worth, in years, the requirement for `pyarrow` has been : `pyarrow` or `pyarrow>=x.y.z`

**Comment by mboisson** [comment]
I see there is a PyArrow 16.1 on PyPi

**Comment by ccoulombe** [comment]
Oops, missed this one but it poses the question : if the user tries this version and install the dummy which tell him to load a module we do not have, it would raise an eyebrow.

I think then this would mean that either the user can use another version, or we do need to install that version of Arrow and then add the dummy ?

**Pull request merged**